torch_geometric/nn/conv/hgt_conv.py

```diff
@@ -139,38 +174,211 @@ def forward(
+            }
+            # compute V
+            v_list = pyg_lib.ops.grouped_matmul(inputs=xs, others=v_wts,
+                                                biases=v_biases)
+            v_dict = {
+                node_type: v_list[i].view(-1, H, D)
+                for i, node_type in enumerate(self.node_types)
+            }
+        else:
+            k = pyg_lib.ops.segment_matmul(inputs=x, ptr=ptr, other=k_wt,
+                                           bias=k_bias)
+            k_dict = {
+                node_type: k[ptr[i]:ptr[i + 1]].view(-1, H, D)
+                for i, node_type in enumerate(self.node_types)
+            }
+            q = pyg_lib.ops.segment_matmul(inputs=x, ptr=ptr, other=q_wt,
+                                           bias=q_bias)
+            q_dict = {
+                node_type: q[ptr[i]:ptr[i + 1]].view(-1, H, D)
+                for i, node_type in enumerate(self.node_types)
+            }
+            v = pyg_lib.ops.segment_matmul(inputs=x, ptr=ptr, other=v_wt,
+                                           bias=v_bias)
+            v_dict = {
+                node_type: v[ptr[i]:ptr[i + 1]].view(-1, H, D)
+                for i, node_type in enumerate(self.node_types)
+            }
+
+        # parallelize over edge-types
+        src_types = [edge_type[0] for edge_type in self.edge_types]
+        a_rels = [
+            self.a_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+        m_rels = [
+            self.m_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+        if self.use_gmm:
+            k_ins = [
+                k_dict[src_type].transpose(0, 1) for src_type in src_types
+            ]
+            v_ins = [
+                v_dict[src_type].transpose(0, 1) for src_type in src_types
+            ]
+            k_outs = [
+                k_o_i.transpose(1, 0)
+                for k_o_i in pyg_lib.ops.grouped_matmul(k_ins, a_rels)
+            ]
+            v_outs = [
+                v_o_i.transpose(1, 0)
+                for v_o_i in pyg_lib.ops.grouped_matmul(v_ins, m_rels)
+            ]
+            node_slices = {}
+            cumsum = 0
+            for i, n_type in enumerate(self.node_types):
+                num_nodes = k_outs[i].shape[0]
+                node_slices[n_type] = (cumsum, cumsum + num_nodes)
+                cumsum += num_nodes
+            k_out = torch.cat(k_outs)
+            v_out = torch.cat(v_outs)
+        else:
+            k_ins = []
+            v_ins = []
+            count = 0
+            trans_ptr = [count]
+            for src_type in src_types:
+                k_src = k_dict[src_type]
+                v_src = v_dict[src_type]
+                k_ins.append(k_src.view(-1, D))
+                v_ins.append(v_src.view(-1, D))
+                for i in range(H):
+                    count += k_src.size(0)
+                    trans_ptr.append(count)
+            trans_ptr = torch.tensor(trans_ptr)
```

Comment:
device?

Reply:
> For best performance, given as a CPU tensor.
don't we recommend cpu tensors for segment matmul?

Reply:
Oh, you are right. Thanks for clarifying.